ably/realtime_conn_spec_integration_test.go

```diff
@@ -1684,18 +1694,214 @@ func TestRealtimeConn_RTN15h2_Success(t *testing.T) {
+	t.Parallel()
+	app, restClient := ablytest.NewREST()
+	defer safeclose(t, app)
+	recorder := NewMessageRecorder()
+
+	authCallbackTokens := []string{}
+	// Returns token that expires after 3 seconds causing disconnect every 3 seconds
+	authCallback := func(ctx context.Context, tp ably.TokenParams) (ably.Tokener, error) {
+		token, err := restClient.Auth.RequestToken(context.Background(), &ably.TokenParams{TTL: 3000})
+		authCallbackTokens = append(authCallbackTokens, token.Token)
+		return token, err
+	}
+
+	realtime, err := ably.NewRealtime(
+		ably.WithAutoConnect(false),
+		ably.WithDial(recorder.Dial),
+		ably.WithEnvironment(ablytest.Environment),
+		ably.WithAuthCallback(authCallback))
+
+	assert.NoError(t, err)
+	defer realtime.Close()
+
+	err = ablytest.Wait(ablytest.ConnWaiter(realtime, realtime.Connect, ably.ConnectionEventConnected), nil)
+	assert.NoError(t, err)
+
+	changes := make(ably.ConnStateChanges, 2)
+	off := realtime.Connection.OnAll(changes.Receive)
+	defer off()
+	var state ably.ConnectionStateChange
+
+	for i := 0; i < 3; i++ {
+		ablytest.Soon.Recv(t, &state, changes, t.Fatalf)
+		assert.Equal(t, ably.ConnectionEventDisconnected, state.Event)
+		assert.Equal(t, ably.ConnectionStateDisconnected, state.Current)
+		assert.Error(t, state.Reason)
+		assert.Equal(t, 401, state.Reason.StatusCode)
+		assert.Equal(t, 40142, int(state.Reason.Code))
+		assert.ErrorContains(t, state.Reason, "token expired")
+
+		ablytest.Soon.Recv(t, &state, changes, t.Fatalf)
+		assert.Equal(t, ably.ConnectionEventConnecting, state.Event)
+		ablytest.Soon.Recv(t, &state, changes, t.Fatalf)
+		assert.Equal(t, ably.ConnectionEventConnected, state.Event)
+		assert.Nil(t, state.Reason)
+		assert.Equal(t, ably.ConnectionStateConnected, realtime.Connection.State())
+	}
+	ablytest.Instantly.NoRecv(t, nil, changes, t.Fatalf)
+	assert.True(t, ablytest.Instantly.IsTrue(recorder.CheckIfReceived(ably.ActionDisconnected, 3)))
+
+	tokens := []string{}
+	assert.Len(t, recorder.URLs(), 4) // 4 connect attempts made in total, disconnect received after each one
+	for _, url := range recorder.URLs() {
+		tokens = append(tokens, url.Query().Get("access_token"))
+	}
+	assert.Len(t, tokens, 4) // 4 tokens explicitly requested and supplied for every attempt
+	assertUnique(t, tokens)  // Make sure all tokens are unique for every connection attempt
+	assert.ElementsMatch(t, authCallbackTokens, tokens)
+}
+
+func TestRealtimeConn_RTN22_RTC8_Integration_ServerInitiatedAuth(t *testing.T) {
+	app, restClient := ablytest.NewREST()
+	defer safeclose(t, app)
+
+	recorder := NewMessageRecorder()
+	authCallbackTokens := []string{}
+
+	// Server sends AUTH message 30 seconds before token expiry.
+	// So sending client token with expiry of 33 seconds, server will send AUTH msg after 3 seconds.
+	authCallback := func(ctx context.Context, tp ably.TokenParams) (ably.Tokener, error) {
+		tokenExpiry := 33000
```

Comment:
I would make `tokenExpiry` a  `time.Duration`

Reply:
Ttl accepts int value in terms of milliseconds. Time.duration is measured in nanoseconds

Reply:
normally one one do something like
```
tokenExpery := 33 * time.Second
token, err := restClient.Auth.RequestToken(context.Background(), &ably.TokenParams{TTL: tokenExpiry.Milliseconds()})
```

Reply:
Ohkay, let me check on this one

Reply:
updatedhttps://github.com/ably/ably-go/pull/661/commits/edce32f870634997fffa86c28687e644baecc8b3